history/history.c

```diff
@@ -0,0 +1,1079 @@
+
+    if (pids) {
+        /* count installed packages */
+        rpmdbMatchIterator mi =
+            rpmtsInitIterator(ts, RPMDBI_PACKAGES, NULL, 0);
+        while ((h = rpmdbNextIterator(mi))) {
+            count++;
+        }
+        rpmdbFreeIterator(mi);
+        ids = (int *)calloc(count, sizeof(int));
+    }
+
+    char *sql = "CREATE TABLE IF NOT EXISTS "
+        "rpms(Id INTEGER PRIMARY KEY AUTOINCREMENT, nevra TEXT);";
+    rc = sqlite3_exec(db, sql, 0, 0, NULL);
+    check_db_rc(db, rc);
+
+    mi = rpmtsInitIterator(ts, RPMDBI_PACKAGES, NULL, 0);
+    while ((h = rpmdbNextIterator(mi))) {
+        nevra = headerGetAsString(h, RPMTAG_NEVRA);
+        rc = db_add_nevra(db, nevra, ids ? &ids[i++] : NULL);
+        check_db_rc(db, rc);
+        safe_free(nevra);
+    }
+    rpmdbFreeIterator(mi);
+
+    if (pids && pcount) {
+        sort_array(ids, count);
+
+        *pids = ids;
+        *pcount = count;
+    }
+    return rc;
+error:
+    safe_free(nevra);
+    if (mi)
+        rpmdbFreeIterator(mi);
+    if (ids)
+        free(ids);
+    return rc;
+}
+
+/* Helper function to add a transaction entry to the db.
+   Does not add transaction items */
+static
+int db_add_transaction(sqlite3 *db, int *ptrans_id,
+                       const char *cmdline, time_t timestamp,
+                       const char *cookie, int type)
+{
+    int rc = 0, ret;
+    sqlite3_stmt *res = NULL;
+
+    rc = sqlite3_prepare_v2(db,
+        "INSERT INTO transactions(cmdline, cookie, timestamp, type) VALUES (?, ?, ?, ?);",
+        -1, &res, 0);
+    check_db_rc(db, rc);
+
+    sqlite3_bind_text(res, 1, cmdline, -1, NULL);
+    sqlite3_bind_text(res, 2, cookie, -1, NULL);
+    sqlite3_bind_int(res, 3, timestamp);
+    sqlite3_bind_int(res, 4, type);
+
+    ret = sqlite3_step(res);
+    check_cond(ret == SQLITE_DONE);
+    *ptrans_id = sqlite3_last_insert_rowid(db);
+
+error:
+    if (res)
+        sqlite3_finalize(res);
+    return rc;
+}
+
+/* Helper function to add transaction items */
+static
+int db_add_trans_items(sqlite3 *db, int trans_id, int type,
+                       int *rpm_ids, int rpm_count)
+{
+    int rc = 0, ret;
+    int i;
+    sqlite3_stmt *res = NULL;
+
+    for (i = 0; i < rpm_count; i++) {
+        rc = sqlite3_prepare_v2(db,
+            "INSERT INTO trans_items(trans_id, type, rpm_id) VALUES (?, ?, ?);",
+            -1, &res, 0);
+        check_db_rc(db, rc);
+
+        sqlite3_bind_int(res, 1, trans_id);
+        sqlite3_bind_int(res, 2, type);
+        sqlite3_bind_int(res, 3, rpm_ids[i]);
+
+        ret = sqlite3_step(res);
+        check_cond(ret == SQLITE_DONE);
+
+        sqlite3_finalize(res); res = NULL;
+    }
+error:
+    if (res)
+        sqlite3_finalize(res);
+    return rc;
+}
+
+
+/* Helper to set the ctx cookie, free'ing the old one if needed */
+static
+void history_set_cookie(struct history_ctx *ctx, const char *cookie)
+{
+    safe_free(ctx->cookie);
+    ctx->cookie = strdup(cookie);
```

Comment:
TdnfAllocateMemory instead of strdup if it does the needful

Reply:
This API uses another coding style than the main TDNF code.